content/blog/2022-06-28-serving-models-with-mlem.md

```diff
@@ -0,0 +1,277 @@
+commentsUrl: https://discuss.dvc.org/t/syncing-data-to-aws-s3/1192
+tags:
+  - MLOps
+  - DVC
+  - Git
+  - MLEM
+  - Collaboration
+---
+
+Training a machine learning model is only one step in the process of getting
+something useful out to the end users. The goal of machine learning projects to
+make an accurate model that provides predictions based on the data received from
+a production environment. That could mean making predictions on data coming from
+an API or do some batch predictions.
+
+Either way, you'll need to save your trained and validated model in format
+that's consumable by other systems. That's why we'll be covering how to serve
+models through an endpoint with [MLEM](https://mlem.ai/). We'll start by
+training a simple model. You can get the repo we're working with
+[here](https://github.com/iterative/stale-model-example/tree/mlem-serve).
+
+## Train a model
+
+There are instructions in the project
+[README](https://github.com/iterative/stale-model-example/tree/mlem-serve#readme)
+on how to get everything you need installed and running. Make sure you run a few
+experiments with:
+
+```dvc
+$ dvc exp run
+```
+
+That way you can see what it's like actually training a model with DVC. We use
+some bicycle data to make a prediction model that helps end users know how many
+bikes to place around an area of the city. After you've run a few experiments,
+take a look at the metrics table:
+
+```dvc
+$ dvc exp show --drop 'Created|./clf|./data|src/evaluate.py|./src/train.py|./models/model.pkl|./data/train.pkl'
+```
+
+_NOTE:_ We're dropping a lot of columns to simplify the data shown in the table.
+
+```dvctable
+  neutral:**Experiment**                metric:**avg_prec**   metric:**roc_auc**   param:**train.seed**   param:**train.n_est**   param:**train.min_split**
+ ───────────────────────────────────────────────────────────────────────────────────────────
+  workspace                  0.72417   0.38926   20210428     250           75
+  mlem-serve                 0.76681   0.38867   20210428     300           75
+  ├── c8d952b [exp-4b026]    0.72417   0.38926   20210428     250           75
+  ├── da78e7c [exp-7a2e2]    0.74454    0.5528   20210428     450           75
+  ├── b39470f [exp-5ebf6]    0.42526   0.82722   20210428     400           28
+  ├── 03c6d57 [exp-f8c3a]    0.74454    0.5528   20210428     400           75
+  ├── 75b1b3c [exp-2ba75]    0.67913   0.38207   20210428     350           75
+  └── 40b0075 [exp-e0dc1]    0.76681   0.38867   20210428     300           75
+ ───────────────────────────────────────────────────────────────────────────────────────────
+```
+
+Of course you would run many more experiments in order to determine the best
+model for production, but in this example, the metrics for `exp-5ebf6` look
+pretty good. So we'll take this model and get it ready to serve to production.
+
+## Save the model
+
+First, we need to install the `MLEM` package with:
+
+```dvc
+$ pip install 'mlem[fastapi]'
```

Comment:
Could we do this first though, install MLEM, plug it in the existing code, and THEN `dvc exp run` to train and select a model. (Idea being: Start with MLEM since it's a MLEM post.)

Reply:
I made a pretty big update to the whole structure and moved the DVC stuff into an info box. I think that should clear the way for MLEM to stay the focus.